iris-ugrid/ugrid_cf_reader.py

```diff
@@ -0,0 +1,231 @@
+                    # from dims of a non-optional connectivity variable.
+                    edges_varname = mesh_var.edge_node_connectivity
+                    edges_var = dataset.variables[edges_varname]
+                    edges_dim_name = edges_var.dimensions[0]
+                meshdims_map[edges_dim_name] = (mesh, "edge")
+            if mesh.nodes is not None:
+                # Work out name of nodes dimension and record it.
+                # Get it from a non-optional coordinate variable.
+                nodes_varname = mesh_var.node_coordinates.split()[0]
+                nodes_var = dataset.variables[nodes_varname]
+                nodes_dim_name = nodes_var.dimensions[0]
+                meshdims_map[nodes_dim_name] = (mesh, "node")
+        self.meshdims_map = meshdims_map
+
+        # Create a CFReader object which skips the UGRID-related variables.
+        kwargs["exclude_var_names"] = exclude_vars
+        self.cfreader = CFReader(self.dataset, *args, **kwargs)
+
+    def complete_ugrid_cube(self, cube):
+        """
+        Add the ".ugrid" property to a cube loaded with the `self.cfreader`.
+        We identify the unstructured-grid dimension of the cube (if any), and
+        attach a suitable CubeUgrid object, linking the cube mesh dimension to
+        an element-type (aka "mesh_location") of a mesh.
+        """
+        # Set a 'cube.ugrid' property.
+        data_var = self.dataset.variables[cube.var_name]
+        meshes_info = [
+            (i_dim, self.meshdims_map.get(dim_name))
+            for i_dim, dim_name in enumerate(data_var.dimensions)
+            if dim_name in self.meshdims_map
+        ]
+        if len(meshes_info) > 1:
+            msg = "Cube maps more than one mesh dimension: {}"
+            raise ValueError(msg.format(meshes_info))
+        if meshes_info:
+            i_dim, (mesh, mesh_location) = meshes_info[0]
+            mesh_var = self.dataset.variables[mesh.mesh_name]
+
+            topology_dimension = mesh_var.getncattr("topology_dimension")
+            node_coordinates = []
+            for node_var_name in mesh_var.getncattr("node_coordinates").split(
+                " "
+            ):
+                node_var = self.dataset.variables[node_var_name]
+                name = (
+                    getattr(node_var, "standard_name", None)
+                    or getattr(node_var, "long_name", None)
+                    or node_var_name
+                )
+                node_coordinates.append(name)
+
+            cube.ugrid = CubeUgrid(
+                cube_dim=i_dim,
+                grid=mesh,
+                mesh_location=mesh_location,
+                topology_dimension=topology_dimension,
+                node_coordinates=sorted(node_coordinates),
+            )
+        else:
+            # Add an empty 'cube.ugrid' to all cubes otherwise.
+            cube.ugrid = None
+        return
+
+    def __del__(self):
```

Comment:
Can we rely on the `__del__` method just being in `CFReader`, thereby not needing the same method in `UGridCFReader`? Can test by checking that running `__del__` on the `UGridCFReader().cfreader` actually closes the dataset, and also by seeing if [iris.tests.integration.test_climatology](https://github.com/SciTools/iris/blob/master/lib/iris/tests/integration/test_climatology.py) still runs (known to have trouble if there is an open dataset).